Per layer progress

ilopezluna opened the pull request:
Update pull progress to track progress per layer

Changes:
- Create a `ProgressReporter` per layer instead of one for the whole model
- Add layer field in `ProgressReporter` to track individual layer progress
- Increase `MinBytesForUpdate` from 1MB to 5MB to reduce progress update frequency
- Use buffered channel (size 100) in `ProgressReporter` to reduce backpressure

Note: Push progress reporting remains unchanged and can be addressed in a follow-up PR.

xenoscopic commented:
I don't know if it's relevant to this PR, but one of the things I've been a little concerned about is the synchronization overhead of the progress updates given that the channel is blocking.  I drafted up a commit here for a proposal that would drop intermediate updates if they're not serviced fast enough:

https://github.com/docker/model-distribution/commit/d35eb0511280c0fe96c195769082810f9142725d

My concern is that if every `Read` system call requires a progress update to be pushed to a client (which may be remote and/or high-latency), then the overhead becomes much higher for each `Read` call.  My main concern was running `docker model pull` requests against a cloud engine, where progress updates would have to be relayed back to the client on the user's laptop.  Each `Read` call would then incur additional latency.  Maybe the kernel's TCP buffers are already handling this a bit?  I'm not sure.

Anyway, I just thought I'd bring it up in case it spurred any thoughts w.r.t. this PR.

ilopezluna commented:
> d35eb05

Thanks for raising this, @xenoscopic. I wasn’t aware of this issue, but what you mentioned makes total sense. I’ll look into it.

ilopezluna commented:
> I don't know if it's relevant to this PR, but one of the things I've been a little concerned about is the synchronization overhead of the progress updates given that the channel is blocking. I drafted up a commit here for a proposal that would drop intermediate updates if they're not serviced fast enough:
> 
> [d35eb05](https://github.com/docker/model-distribution/commit/d35eb0511280c0fe96c195769082810f9142725d)
> 
> My concern is that if every `Read` system call requires a progress update to be pushed to a client (which may be remote and/or high-latency), then the overhead becomes much higher for each `Read` call. My main concern was running `docker model pull` requests against a cloud engine, where progress updates would have to be relayed back to the client on the user's laptop. Each `Read` call would then incur additional latency. Maybe the kernel's TCP buffers are already handling this a bit? I'm not sure.
> 
> Anyway, I just thought I'd bring it up in case it spurred any thoughts w.r.t. this PR.

@xenoscopic in [this commit](https://github.com/docker/model-distribution/pull/80/commits/b3cadcec77f5f2f6361506b53eec8c7dd79347af) I addressed the potential backpressure issue.
I'm now using a buffered channel (with a size of 100) and have increased the minimum size to trigger an update to 5 MB.
I also added a test called "unbuffered vs buffered channel" to verify the performance improvement, and I'm consistently seeing a 2x speedup locally.
Please take a look when you have time and let me know wdyt 🙏

xenoscopic commented on internal/progress/reporter.go:
```diff
@@ -35,12 +51,13 @@ func PushMsg(update v1.Update) string {
 	return &Reporter{
 		out:      w,
-		progress: make(chan v1.Update),
+		progress: make(chan v1.Update, 100),
```
I don't think there's any value in buffering this beyond 1 element - any buffer size will eventually start transmitting backpressure if there's no mechanism for dropping intermediate updates that aren't serviced in a certain period of time (i.e. if the producer is producing faster than the consumer is consuming, you'll always fill the buffer in time, and then you'll be back to blocking again, and the only way to avoid that is to drop intermediate updates).  I think the only pattern where buffering > 1 really makes sense is when using the channel as a semaphore.  I would buffer 1 element and add either a `default:` case or a timeout case on the channel write.

ilopezluna replied:
Thanks for the explanation, I didn’t properly understand it before.. 
I have added your suggestion of dropping intermediate updates here: [a818eab](https://github.com/docker/model-distribution/pull/80/commits/a818eabd9a64290e74051fc4c63667bd1097d0e4)
